Improve performance

trowski opened the pull request:
## Proposal

Improved parser speed significantly by avoiding unnecessary function calls and using regular expressions over `strspn`

### Backward Incompatible Changes

If the constants were considered public API I will have to re-add them even though they'll be unused.

trowski committed:
Use regexp and other changes for improved performance

nyamsprod commented:
@trowski thanks for the contributions I'll have to review the submitted changes first before merging. 

One question thought are you sure regular expressions are indeed faster than `strspn` ? I was under the impression of the opposite but I may be wrong. for me `strspn` is like `strpos` which is faster that regular expressions

nyamsprod commented on src/Parser.php:
```diff
@@ -71,9 +69,10 @@ public function isScheme(string $scheme): bool
     {
         return '' === $host
+            || $this->isRegisteredName($host)
             || filter_var($host, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)
```
FWIW ... a IPv4 host is a "special registered name" so for a URI parser point of view there's no necessity to distinguish between a registered name and an IPv4 host. this check can be safely remove

nyamsprod commented on src/Parser.php:
```diff
@@ -164,23 +166,23 @@ public function parse(string $uri): array
 
         //The URI is made of the fragment only
-        if ('#' == $first_char) {
+        if ('#' === $first_char) {
```
while I understand the use of `===` in this case that make no much difference because `$first_char` will always be a `string`. The comment will be the same for each comparison done against `$first_char`

kelunik replied:
This is rather a generic code clean-up than a performance change. `==` usually shouldn't be used, at it can lead to subtle bugs.

nyamsprod replied:
agreed, it's more a generic code clean up than a performance change. `===` is already used in critical place... but I have no objection to make it a requirement in the codebase

nyamsprod commented on src/Parser.php:
```diff
@@ -317,7 +319,7 @@ protected function filterHost($host)
     protected function isIpv6Host(string $ipv6): bool
     {
-        if ('][' !== substr($ipv6.$ipv6, strlen($ipv6) - 1, 2)) {
+        if (2 > strlen($ipv6) || '[' !== $ipv6[0] || ']' !== substr($ipv6, -1)) {
```
the `$ipv6` string can not be the empty string.. the check is done prior to calling this method so you don't need to check its length.

trowski replied:
Since the method is protected, technically no assumptions should be made about the input here.

nyamsprod replied:
yes the method is protected (should have been private .. and the whole class final ... but that's another topic for another day)  meaning as a maintainer I do know when and why it will be used so I'm in the camp that I can make assumptions. If it was a public API method I'll second your change because there I have no way to control the user input.

kelunik replied:
Instead of a function call to `strlen`, this could use `($ipv6[0] ?? '') !== ...`. On the other hand, `\strlen` is optimized to its own opcode anyway, but only if it's prefixed with `\`.

nyamsprod commented on src/Parser.php:
```diff
@@ -87,8 +86,11 @@ public function isHost(string $host): bool
+            return true;
+        }
+
+        return (bool) filter_var($port, FILTER_VALIDATE_INT, ['options' => ['min_range' => 1, 'max_range' => 65535]]);
```
why casting to boolean ? the conditions states the port must be at least 1 so any valid value will return a positive integer thus `true` ?

trowski replied:
With strict types, a `TypeError` is thrown if the value returned isn't explicitly a boolean.

trowski commented:
Like so many things with performance, the answer is "it depends." `strspn` can be faster with small dictionaries or short strings. The large dictionaries here seem to drag down performance. Additionally regular expressions allow checks that would otherwise require further function calls and/or string manipulation.

[Here is a Blackfire comparison](https://blackfire.io/profiles/compare/31714ce5-3c8a-4adf-93ee-2124b90a7bf6/graph) of the benchmark @kelunik added to `league/uri-schemes` before and after this PR. Most of the time saving is in `isRegisteredName`, but `isScheme` is interesting. The primary change there was switching from `strspn` to a regular expression on a relatively short string. The execution time used by `isScheme` was reduced 56%.

kelunik commented:
@trowski I think `isScheme` could actually be further optimized by using a `static $knownSchemes = ["http" => true, ...]` map that includes common schemes such as `http` and `https`. There's no need for include all variations like `HtTP`, because the callback to the `preg_match` will cover these.

kelunik commented on src/Parser.php:
```diff
@@ -26,13 +26,15 @@
-    const LABEL_VALID_STARTING_CHARS = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
+    const DOMAIN_LABEL_REGEXP = '/^[a-z0-9\-_]{1,63}$/i';
+
+    const LABEL_REGEXP = '/^[a-z0-9!$&\-\'\(\)*+,;=_~]+$/i';
```
This should be `HOST_LABEL_REGEXP` to avoid confusion with `DOMAIN_LABEL_REGEXP`.

kelunik commented:
@trowski The `*_REGEXP` constants are all used in a single method only, I'd just convert these to `static $pattern = '...'` variables I think, which avoid exposing them as public API.

trowski commented:
Pushed a new commit based on feedback.

kelunik commented on src/Parser.php:
```diff
@@ -363,15 +358,26 @@ protected function isIpv6Host(string $ipv6): bool
         }
 
-        $labels = array_map([$this, 'toAscii'], explode('.', $host));
+        if (preg_match($pattern, $host)) {
```
This misses the overall length check now. The regex just checks the number of labels and their length.

trowski replied:
Was there an overall length check? Seems to me only 127 * 63 was being enforced, and still is by the regex.

kelunik commented on src/Parser.php:
```diff
@@ -363,15 +358,26 @@ protected function isIpv6Host(string $ipv6): bool
+            return false;
+        }
+
+        $labels = array_filter(array_map([$this, 'toAscii'], $labels));
```
This is wrong as it filters `0`, which is a valid label name. Also, it can be optimized to use a single `idn_to_ascii` call. `toAscii` can probably be deprecated then.

trowski replied:
Yep, good call.

I also didn't know that `idn_to_ascii` could be done over the entire string, that will simplify things a lot.

kelunik commented on src/Parser.php:
```diff
@@ -363,15 +358,26 @@ protected function isIpv6Host(string $ipv6): bool
      */
     protected function isRegisteredName(string $host): bool
     {
+        static $pattern = '/^(?:[a-z0-9\-_]{1,63}\.){0,126}[a-z0-9\-_]{1,63}$/i';
```
As far as I know `-` isn't allowed as leading character for a DNS name and `_` isn't allowed at all.

trowski replied:
Not in a registered name, as far as I know you can use it in a DNS name, so subdomains could.

nyamsprod replied:
`_` and `-` are allowed because i had to resolve [the following bug](https://github.com/thephpleague/uri/issues/107). A DNS name is a registered name but a registered name is not always a DNS name

kelunik replied:
Right, but then also `~` and a few more characters (`sub-delims`) needs to be allowed and the length requirements as well as the IDNA normalization need to be entirely dropped. See https://tools.ietf.org/html/rfc3986#section-3.2.2 + https://tools.ietf.org/html/rfc3986#appendix-A.

>     reg-name    = *( unreserved / pct-encoded / sub-delims )

>     unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"

>     sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
>                   / "*" / "+" / "," / ";" / "="

trowski committed:
Conform better to RFC3986

The RFC states that names must begin or end with an alphanumeric character, however the regex here also allows other gen_delims and sub_delims, except -, as that is not allowed in DNS and fails when calling idn_to_ascii.

Modifed the test containing com-, added some others with dashes that are expected fails, as well as adding a few other with gen_delims and sub_delims.

trowski committed:
Avoid some array_pop() calls and superfluous compares

staabm commented on src/Parser.php:
```diff
@@ -54,9 +44,9 @@ class Parser
-        return '' === $scheme
-            || (strlen($scheme) === strspn($scheme, self::SCHEME_VALID_CHARS)
-                && false !== strpos(self::SCHEME_VALID_STARTING_CHARS, $scheme[0]));
+        static $pattern = '/^[a-z][a-z\+\.\-]*$/i';
```
do static vars offer a perf benefit over class constants?

kelunik replied:
This has just been done, so they are not public API.

staabm replied:
ok so actually this is a BC break.

kelunik replied:
With constants it's a BC break, too. But we can re-add the old constants and mark them as deprecated.

nyamsprod replied:
the constants can be re-introduced and left as is to avoid any BC break if they are not used anymore. if it was a PHP7.1+ library they would have been marked as private. I did not document them for that reason on the documentation website. They were meant to be internal to the class

kelunik replied:
@nyamsprod Better mark such places with `/** @internal */` next time, so people can be sure it's internal and not public API, largely avoids this issue.

Making the class `final` and all methods `private` also help reducing such things being BC breaks.

nyamsprod replied:
@kelunik agreed like I said this whole class should have been final with private methods only. It will be like that for the next major release.

kelunik commented on src/Parser.php:
```diff
@@ -363,18 +358,37 @@ protected function isIpv6Host(string $ipv6): bool
-            $host = substr($host, 0, -1);
+        // Note that unreserved is purposely missing . and -
+        // . is used to separate labels
+        // - cannot start or end a label
```
There's no reason for this. The RFC allows any sequence, just DNS has that restriction, see https://tools.ietf.org/html/rfc3986#section-3.2.2

>     reg-name    = *( unreserved / pct-encoded / sub-delims )

kelunik commented on src/Parser.php:
```diff
@@ -363,18 +358,37 @@ protected function isIpv6Host(string $ipv6): bool
+            return false;
+        }
+
+        return (bool) preg_match($pattern, $host);
```
There's no need for this check after `idn_to_ascii`, as that already does all validation.

kelunik commented on src/Parser.php:
```diff
@@ -363,18 +358,37 @@ protected function isIpv6Host(string $ipv6): bool
-        return 127 > count($labels)
-            && 253 > strlen(implode('.', $labels))
-            && $labels === array_filter($labels, [$this, 'isHostLabel']);
+        $host = idn_to_ascii($host, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
```
It seems like IDNs are not covered by the defined `reg-name` syntax in https://tools.ietf.org/html/rfc3986#section-3.2.2

>     reg-name    = *( unreserved / pct-encoded / sub-delims )

This means these hosts are strictly coupled to DNS, which means we can apply all DNS restrictions on IDN domains, but not on other host names.

nyamsprod replied:
IDN domains are defined in rfc3987 so yes they do not cover all the reg name they only cover i18n cases

kelunik commented on src/Parser.php:
```diff
@@ -387,18 +401,18 @@ protected function isRegisteredName(string $host): bool
     {
-        if (false !== strpos($label, '%')) {
-            $label = rawurldecode($label);
-        }
```
Let's keep the decoding here for BC purposes, the method is now unused anyway.

kelunik commented on src/Parser.php:
```diff
@@ -363,18 +358,37 @@ protected function isIpv6Host(string $ipv6): bool
     }
 
     /**
+     * @deprecated Will be removed in v2.0
```
@nyamsprod Should we trigger deprecation notices using `trigger_error` or not?

nyamsprod replied:
@kelunik this method is internal so unless someone does extends this class (which I highly doubt) I don't see the advantage of using `trigger_error`. Marking it deprecated is enough for me

kelunik replied:
@nyamsprod If it's intended that way, that's a perfect reason to use `trigger_error`. As we don't use it anymore, any users can be alerted that it will be made private and the complete class be made final in a future release.

nyamsprod replied:
fine by me

trowski replied:
Added `trigger_error` to these methods.

kelunik commented on src/Parser.php:
```diff
@@ -417,9 +431,16 @@ protected function toAscii(string $label)
+                (?<unreserved> [a-z0-9\_\~])
+                (?<gen_delims> [:\/?#\[\]@])
+                (?<sub_delims> [!$&\'()*+,;=])
+                (?<label> (?:(?&unreserved)(?:(?&unreserved)|(?&gen_delims)|(?&sub_delims)|\-){0,61})?(?&unreserved))
```
There's no length requirement in the RFC for labels as far as I can tell. Actually, there isn't even a definition of label at all, it just refers to the labels of DNS if DNS is used.

kelunik commented on src/Parser.php:
```diff
@@ -527,9 +550,7 @@ protected function parsePathQueryAndFragment(string $uri): array
-        $parts = explode(':', $uri, 2);
-        $remaining_uri = array_pop($parts);
-        $scheme = array_shift($parts);
+        list($scheme, $remaining_uri) = explode(':', $uri, 2) + [1 => null];
```
This is wrong, because it assigns `$scheme` now if there's no colon instead of `$remaining_uri`.

nyamsprod replied:
@trowski could you revert this change please ?

nyamsprod replied:
maybe use the other construct
```php
$parts = explode(':', $uri, 2);
$remaining_uri = $parts[1] ?? $parts[0];
$scheme = isset($parts[1]) ? $parts[0] : null;
```

trowski replied:
Fixed.

kelunik commented on tests/ParserTest.php:
```diff
@@ -684,8 +684,11 @@ public function validHostProvider()
+            'RFC3986 registered name (2)' => ['www._fußball.com', true],
+            'Sub-domain beginning with underscore' => ['_tcp.example.com', true],
             'Host with urlencoded label' => ['b%C3%A9b%C3%A9.be', true],
+            'Host with delims' => ['tes:/?#+&[]@t.com', true],
```
`/`, `#`, `[`, `]` and `@` aren't `sub-delims`, they're `gen-delims` and are not allowed in the host. I guess this is currently parsed as path `/` with an empty query and the fragment `+&[]@t.com`.

nyamsprod replied:
A better test would be to see if `tot...toto.com` returns `true`. The parser should let it pass. The URI objects I don't know 🤔

nyamsprod commented on src/Parser.php:
```diff
@@ -435,15 +456,17 @@ protected function isHostLabel($label): bool
         }
 
-        if (!$this->isPort($port)) {
+        $result = filter_var($port, FILTER_VALIDATE_INT, ['options' => ['min_range' => 1, 'max_range' => 65535]]);
```
referring to issue #15 we can simplify this code using a simple regex to check that the port contains only digit. I believe this means that port `0` is accepted too but negative port are still forbidden

trowski committed:
Fix reg-name regex; allow any positive number for port

trowski commented:
Pushed another commit based on our discussions.

Open issues:
- Host labels are limited to 63 characters.
- Host labels cannot be empty. For example, `www.te..st.com` will fail.
- Host names are limited to 127 labels.
- Port numbers now can be any positive number. What about negative numbers?

My initial thought is that the above limitations make sense, and any URI in the wild should conform to this, but the RFC does not specify any of the above limitations. Thoughts?

trowski committed:
Restore rawurldecode for BC

nyamsprod commented:
@trowski according to [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2.3) Port numbers only contains digits so ... negative numbers are not supported.

nyamsprod commented:
@trowski for the host issues according to RFC3986 the only restriction is 

> This specification does not mandate a particular registered name
   lookup technology and therefore does not restrict the syntax of reg-
   name beyond what is necessary for interoperability.  Instead, it
   delegates the issue of registered name syntax conformance to the
   operating system of each application performing URI resolution, and
   that operating system decides what it will allow for the purpose of
   host identification.  A URI resolution implementation might use DNS,
   host tables, yellow pages, NetInfo, WINS, or any other system for
   lookup of registered names.  However, a globally scoped naming
   system, such as DNS fully qualified domain names, is necessary for
   URIs intended to have global scope.  URI producers should use names
   that conform to the DNS syntax, even when use of DNS is not
   immediately apparent, and should limit these names to no more than
   255 characters in length.

For me this reads like the hard requirement is **no more than 255 characters in length.** Everything else is in the realm of how far do we take interoperability and correctness. The presence, the length of label is not mentioned in the RFC. 

For instance currently IP futures syntax is not supported by the parser. Because, I never encountered them. Maybe I'm wrong and we should add them 🤔 

@kelunik what do you think ?

kelunik commented:
@nyamsprod That's also a should, which is only there due to DNS interoperability. There's no hard limit, only a recommendation. Either we limit all hosts to DNS or not at all. I'd not limit the host's name at all then.

I don't think it's worth adding support for future IP versions. They're rather unlikely and can always be added later.

nyamsprod commented:
@kelunik agreed the part on future IP versions 👍 . 

>I'd not limit the host's name at all then.

This means that the parser should not fail on `www.te..st.com` too. I'm not against it I just want everyone to be on the same page to have a coherent en predictive parser.

kelunik commented:
@nyamsprod Correct. The only exception are IDN domains, because they're not covered by `reg-name`.

nyamsprod commented:
@trowski could you add a `benchmark/parse.php` file in your PR with the following content ?
```php
<?php

require __DIR__.'/../vendor/autoload.php';

$uri = 'https://uri.thephpleague.com/5.0';
for ($i = 0; $i < 100000; $i++) {
    League\Uri\parse($uri);
}
```

nyamsprod commented:
I'm using [webgrind](https://github.com/jokkedk/webgrind) instead of blackfire.io the interface is not as shiny but the results do give the same kind of information .

kelunik commented:
@nyamsprod How should we proceed here? Should we break BC now or do you want this PR to be keep BC and break it in a future version? Given the changes are rather large and might be considered a BC break anyway, I'd prefer breaking BC now.

nyamsprod commented:
@kelunik since we were able to apply similar changes to the schemes package without BC break I though we could do the same here. In my mind they will serve as last stable release for the v1 series. Meanwhile, I can create a develop branch for the future v2 series.

kelunik commented on src/Parser.php:
```diff
@@ -363,18 +360,31 @@ protected function isIpv6Host(string $ipv6): bool
-        return 127 > count($labels)
-            && 253 > strlen(implode('.', $labels))
-            && $labels === array_filter($labels, [$this, 'isHostLabel']);
+        return (bool) idn_to_ascii($host, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
```
This should have a note that it's fine due to the above regex. Otherwise this would mark `'0'` as invalid.

kelunik commented on src/Parser.php:
```diff
@@ -417,9 +430,22 @@ protected function toAscii(string $label)
         return '' != $label
-            && 63 >= strlen($label)
-            && strlen($label) == strspn($label, self::LABEL_VALID_STARTING_CHARS.'-_~'.self::SUB_DELIMITERS);
+            && preg_match($reg_name, $label);
```
I'd just keep the old behavior for this method TBH.

kelunik replied:
Ping @trowski.

kelunik commented:
@nyamsprod Agreed, it's easy enough to support that. I think we make the accepted ranges (e.g. the port ranges) only wider, everything that was previously supported is still supported. Well, except for `fußball-.de`, which is invalid now. Not sure whether that counts as BC break, because it's an invalid IDN, but there was even a test for it.

@trowski Could you restore the constants and mark them as deprecated then? Also, please resolve the conflicts.

nyamsprod commented:
@kelunik I consider correcting the parser behaviour around `fußball-.de` a bug fix. So yeah for me there are no BC break. I've ported some of the changes already on uri-schemes and uri-component so we will wait for @trowski to resolve the pending issues and we can release the 3 packages at once 👍

trowski committed:
Remove length restrictions

trowski committed:
Add parsing benchmark

trowski committed:
Merge branch 'master' into performance-improvements

nyamsprod commented:
@trowski just re-introduce the removed constant to avoid BC break and I'll merget this one :+1:

trowski committed:
Restore constants for BC

trowski commented:
@nyamsprod I think this is ready for merge now. :shipit:

kelunik commented:
@trowski Maybe add a `@deprecated` annotations to the constants.

trowski committed:
Add deprecated annotations

trowski commented:
@kelunik Good idea, done.

kelunik commented on src/Parser.php:
```diff
@@ -26,12 +26,16 @@
+    /** @deprecated Will be removed in v2.0 */
     const LABEL_VALID_STARTING_CHARS = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
 
     const LOCAL_LINK_PREFIX = '1111111010';
```
Same here.

trowski replied:
Skipped it because it was still used, but I think @nyamsprod said they would be made private, so I'll add them there too.

kelunik replied:
@trowski See https://github.com/thephpleague/uri-parser/pull/11#issuecomment-372038760.

kelunik commented:
Please modify the `isIpv6Host` method like it's done [here](https://github.com/thephpleague/uri-schemes/blob/61514cd3db18f481e2343f57def7baedc3e182e8/src/AbstractUri.php#L432-L438).

trowski committed:
Examine IPv6 zone ID

kelunik commented on src/Parser.php:
```diff
@@ -351,13 +352,23 @@ protected function isIpv6Host(string $ipv6): bool
-        return substr($res, 0, 10) === self::LOCAL_LINK_PREFIX;
+        static $local_link_prefix = '1111111010';
+
+        return substr($res, 0, 10) === $local_link_prefix;
```
You can remove the lines 363-371 and just `return substr(inet_pton($ipv6) & $address_block, 0, 2) === $address_block`.

trowski replied:
Oh, awesome. Done.

kelunik commented on src/Parser.php:
```diff
@@ -71,9 +77,9 @@ public function isScheme(string $scheme): bool
-            || filter_var($host, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)
             || $this->isIpv6Host($host)
             || $this->isRegisteredName($host);
+
```
Unnecessary newline.

trowski committed:
Restore old isHostLabel

Method is unused and deprecated anyway.

trowski commented:
Ok, with those minor changes I think now it's good to merge. 😄

kelunik commented:
:tada:
